### PR TITLE
[CatalogPromotion] Apply catalog promotions for product created in catalog promotions scope

### PR DESCRIPTION
**Pull request by ernestWarwas**
| Q               | A
| --------------- | -----
| Branch?         | 1.11 
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no 
| License         | MIT

New products should apply for catalog promotion if falls in catalog promotion scope (eg. when creating a new product in taxon „Clothes” which applies for catalog promotion)

### CODE REVIEW

File: src/Sylius/Behat/Context/Ui/Admin/ProductCreationContext.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    /**
+     * @When /^I create a new simple product ("[^"]+") priced at "(?:€|£|\$)([^"]+)" with ("[^"]+" taxon) in the ("[^"]+" channel)$/
+     */
+    public function iCreateANewVariantPricedAtForProductInTheChannel(
+        string $name,
+        string $price,
+        TaxonInterface $taxon,
+        ChannelInterface $channel
+    ): void {
+        $locale = $channel->getDefaultLocale()->getCode();
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Admin/ProductCreationContext.php:38]:
```suggestion
        $localeCode = $channel->getDefaultLocale()->getCode();
```

File: src/Sylius/Behat/Context/Ui/Admin/ProductVariantsCreationContext.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+final class ProductVariantsCreationContext implements Context
+{
+    public function __construct(private CreatePageInterface $createPage)
+    {
+    }
+
+    /**
+     * @When /^I create a new ("[^"]+") variant priced at "(?:€|£|\$)([^"]+)" for ("[^"]+" product) in the ("[^"]+" channel)$/
+     */
+    public function iCreateANewVariantPricedAtForProductInTheChannel(
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Admin/ProductVariantsCreationContext.php:31]: Why didn't you put this step in `ManagingProductVariantsContext`? IMO it's ok, just asking 😄

Comment (ernestWarwas) [reply src/Sylius/Behat/Context/Ui/Admin/ProductVariantsCreationContext.php:31]: I wanted to make it more reusable. With this approach for me, this context strictly means creating product variants and I think it’s more readable. That’s why I created a new context, I think I should change it also in API contexts. What do you think?

File: src/Sylius/Behat/Context/Api/Admin/ManagingCatalogPromotionsContext.php [hunk 3]
```diff
@@ -23,6 +23,7 @@
 use Sylius\Bundle\CoreBundle\Checker\InForProductScopeVariantChecker;
 use Sylius\Bundle\CoreBundle\Checker\InForTaxonsScopeVariantChecker;
 use Sylius\Bundle\CoreBundle\Checker\InForVariantsScopeVariantChecker;
+use Sylius\Component\Core\Formatter\StringInflector;
```
Comment (AdamKasp) [inline src/Sylius/Behat/Context/Api/Admin/ManagingCatalogPromotionsContext.php:26]: Do we need it?

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductVariantsContext.php [hunk 4]
```diff
@@ -176,4 +189,24 @@ private function updateChannelPricingField(
...
+
+    private function createNewVariantWithPrice(
+        string $name,
+        int $price,
+        ProductInterface $product,
+        ChannelInterface $channel
+    ): void {
+        $this->client->buildCreateRequest();
+        $this->client->addRequestData('product', $this->iriConverter->getIriFromItem($product));
+        $this->client->addRequestData('code', str_replace('"', '', StringInflector::nameToUppercaseCode($name)));
```
Comment (AdamKasp) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductVariantsContext.php:201]: you cen use `StringInflector::nameToCode()`

File: src/Sylius/Bundle/CoreBundle/Listener/ProductVariantCreatedListener.php [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+    public function __invoke(ProductVariantCreated $event): void
+    {
+        /** @var ProductVariantInterface|null $product */
+        $productVariant = $this->productVariantRepository->findOneBy(['code' => $event->code]);
+        if ($productVariant === null) {
+            return;
+        }
+
+        $this->allCatalogPromotionsProcessor->process();
```
Comment (AdamKasp) [inline src/Sylius/Bundle/CoreBundle/Listener/ProductVariantCreatedListener.php:39]: you should use here `CoreBundle/Processor/ProductVariantCatalogPromotionsProcessor.php` service

**Review by AdamKasp — CHANGES_REQUESTED** [review]

**Review by AdamKasp — APPROVED** [review]

**Review by SirDomin — APPROVED** [review]

File: src/Sylius/Behat/Page/Admin/Product/CreateSimpleProductPage.php [hunk 6]
```diff
@@ -250,6 +264,7 @@ protected function getDefinedElements(): array
             'language_tab' => '[data-locale="%locale%"] .title',
             'locale_tab' => '#attributesContainer .menu [data-tab="%localeCode%"]',
             'main_taxon' => '#sylius_product_mainTaxon',
+            'product_taxons' => '#sylius_product_productTaxons',
```
Comment (lchrusciel) [inline src/Sylius/Behat/Page/Admin/Product/CreateSimpleProductPage.php:267]: Can we use test attribute here?

File: features/promotion/applying_catalog_promotions/applying_catalog_promotions_for_product_created_in_catalog_promotions_scope.feature [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    As a Store Owner
+    I want to have catalog promotion applied on new products
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store classifies its products as "Clothes" and "Dishes"
+        And I am logged in as an administrator
+
+    @ui @javascript @no-api
+    Scenario: Applying catalog promotion on newly created simple product
```
Comment (lchrusciel) [inline features/promotion/applying_catalog_promotions/applying_catalog_promotions_for_product_created_in_catalog_promotions_scope.feature:13]: I'm wondering if we should cover product creation without any variant case as well

**Review by lchrusciel — APPROVED** [review]

**Comment by lchrusciel** [comment]
Thank you, @ernestWarwas! :tada:

**Pull request merged**